### PR TITLE
[Update02]Add constraint to all applicable data types in ballerina

### LINKED ISSUE #975: Incoporate OpenAPI type mapping changes
According to the constraint support, https://github.com/ballerina-platform/ballerina-standard-library/issues/2850 need to provide support for the below type validation

- [x] **Primitive type: Integer : (ballerina mapping) int**
<b style="font-weight:normal;" id="docs-internal-guid-6b9c5aaa-7fff-c31d-d728-867bfaa738f2"><div dir="ltr" style="margin-left:0pt;" align="left">

format | keywords | Ballerina Mapping
-- | -- | --
int32 |   | int
int64 |   | int
  | minimum | `@constraint:Int {   minValue: <value>}`
  | maximum | `@constraint:Int {   maxValue: <value>}`
  | exclusiveMinimum | `@constraint:Int {   minValueExclusive: <value>}`
  | exclusiveMaximum | `@constraint:Int {   maxValueExclusive: <value>}`
  | multipleOf | 

</div></b>

- [x] **Primitive type: float (ballerina mapping) : float**
<b style="font-weight:normal;" id="docs-internal-guid-6b9c5aaa-7fff-c31d-d728-867bfaa738f2"><div dir="ltr" style="margin-left:0pt;" align="left">

format | keywords | Ballerina Mapping
-- | -- | --
  | minimum | `@constraint:Float {   minValue: <value>}`
  | maximum | `@constraint:Float {   maxValue: <value>}`
  | exclusiveMinimum | `@constraint:Float {   minValueExclusive: <value>}`
  | exclusiveMaximum | `@constraint:Float {   maxValueExclusive: <value>}`
  | multipleOf | 

</div></b>

- [x] **Primitive type: Double (ballerina mapping) : decimal**
<b style="font-weight:normal;" id="docs-internal-guid-6b9c5aaa-7fff-c31d-d728-867bfaa738f2"><div dir="ltr" style="margin-left:0pt;" align="left">

format | keywords | Ballerina Mapping
-- | -- | --
  | minimum | `@constraint:Number {   minValue: <value>}`
  | maximum | `@constraint:Number {   maxValue: <value>}`
  | exclusiveMinimum | `@constraint:Number {   minValueExclusive: <value>}`
  | exclusiveMaximum | `@constraint:Number {   maxValueExclusive: <value>}`
  | multipleOf | 

</div></b>

- [x] **Primitive type: String (ballerina mapping) : string**
<b style="font-weight:normal;" id="docs-internal-guid-6b9c5aaa-7fff-c31d-d728-867bfaa738f2"><div dir="ltr" style="margin-left:0pt;" align="left">

format | keywords | Ballerina Mapping
-- | -- | --
  | minLength | `@constraint:String {   minLength: <value>}`
  | maxLength | `@constraint:Stringl {   maxLength: <value>}`
(not support from OpenAPI)  | length | `@constraint:Stringl {   length: <value>}`
  | pattern | 
</div></b>

- [x] **Array type: String (ballerina mapping) : array**
<b style="font-weight:normal;" id="docs-internal-guid-6b9c5aaa-7fff-c31d-d728-867bfaa738f2"><div dir="ltr" style="margin-left:0pt;" align="left">

format | keywords | Ballerina Mapping
-- | -- | --
  | minItems | `@constraint:Array {   minLength: <value>}`
  | maxItems | `@constraint:Array {   maxLength: <value>}`
  | uniqueItems | 
</div></b>

- [ ] **Object type: object schema (ballerina mapping) : record type**
<b style="font-weight:normal;" id="docs-internal-guid-6b9c5aaa-7fff-c31d-d728-867bfaa738f2"><div dir="ltr" style="margin-left:0pt;" align="left">

format | keywords | Ballerina Mapping
-- | -- | --
  | minProperties | 
  | maxProperties | 
</div></b>

- [x] Apply this for object schema to record type mapping
- [x] Apply for type definitions 
- [ ] Apply these constraints for functions parameters both remote, resource in service and client generation (ex: query parameters)
- [ ] Add the option for the client generation
- [ ] Update service resource functions
- [ ] Update client remote funtions
- [ ] Update learn page
- [ ] Update help text

### PR DESCRIPTION
**Pull request by lnash94**
Meeting notes with scenarios : https://docs.google.com/document/d/1RwNrZpgU9A7t9gpR_JRTzrmlNWsesC66Od2t_-hhZ3I/edit#heading=h.ywkqvf2gidm2

ex: generated record when it has some validation
```ballerina
import ballerina/constraint;

@constraint:String {maxLength: 23}
public type HobbyItemsString string;

@constraint:Array {maxLength: 5, minLength: 2}
public type Hobby HobbyItemsString[];

@constraint:String {minLength: 7}
public type PersonDetailsItemsString string;

@constraint:Float {maxValue: 445.4}
public type PersonFeeItemsNumber float;

public type Person record {
    Hobby hobby?;
    @constraint:Array {maxLength: 5}
    PersonDetailsItemsString[] salaryDetails?;
    int id;
    PersonFeeItemsNumber[] fee?;
};
```

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

**Comment by ldclakmal** [comment]
Reviewed the constraint package usages in test cases. LGTM.

File: openapi-cli/src/main/java/io/ballerina/openapi/generators/GeneratorUtils.java [hunk 1]
```diff
@@ -461,4 +463,48 @@ public static void addImport(List<ImportDeclarationNode> imports, String module)
...
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public static boolean isAvailableConstraint(Schema<?> value) {
+        if (value instanceof ObjectSchema && value.getProperties() != null) {
+            for (Map.Entry<String, Schema> e : value.getProperties().entrySet()) {
```
Comment (NipunaRanasinghe) [inline openapi-cli/src/main/java/io/ballerina/openapi/generators/GeneratorUtils.java:485]: Cant we simplify the loop as follows?

```suggestion
            for (Schema<?> propertyValue : value.getProperties().values()) {
```

Comment (NipunaRanasinghe) [reply openapi-cli/src/main/java/io/ballerina/openapi/generators/GeneratorUtils.java:485]: You can further improve the logic using stream APIs (`anyMatch(Predicate<? super T> predicate)`)

**Review by NipunaRanasinghe — CHANGES_REQUESTED** [review]

File: openapi-cli/src/main/java/io/ballerina/openapi/generators/schema/TypeGeneratorUtils.java [hunk 2]
```diff
@@ -141,63 +165,239 @@ public static TypeDescriptorNode getNullableType(Schema schema, TypeDescriptorNo
...
+    /**
+     * Generate constraint for numbers : int, float, decimal.
+     */
+    private static AnnotationNode generateNumberConstraint(Schema<?> fieldSchema) {
+        List<String> fields = getNumberAnnotFields(fieldSchema);
+        if (fields.isEmpty()) {
+            return null;
+        }
+        String annotBody = OPEN_BRACE + String.join(COMMA, fields) + CLOSE_BRACE;
+        AnnotationNode annotationNode = createAnnotationNode(CONSTRAINT_INT, annotBody);
```
[the remark this answers is not in the document]
Comment (lnash94) [reply openapi-cli/src/main/java/io/ballerina/openapi/generators/schema/TypeGeneratorUtils.java:276]: This function receives an integer scheme or Number  Schema. There is a condition for checking the number schema. will move this to else block.

**Review by NipunaRanasinghe — APPROVED** [review]
LGTM

**Pull request merged**